common/src/main/java/io/pravega/common/util/SimpleDeque.java

```diff
@@ -0,0 +1,315 @@
+        if (this.items.length >= MAX_CAPACITY) {
+            // Can't allocate an array bigger than this.
+            throw new OutOfMemoryError("Unable to grow SimpleDequeue.");
+        }
+
+        if (newCapacity >= MAX_HALF) {
+            // We double capacity, so if we're already above half of max, then just set it to max.
+            newCapacity = MAX_CAPACITY;
+        } else {
+            // Double capacity.
+            newCapacity = newCapacity * 2;
+        }
+
+        final Object[] newItems = Arrays.copyOf(this.items, newCapacity);
+        boolean flipped = this.tail < this.head || (this.tail == this.head && newItems[this.head] != null);
+        if (flipped) {
+            final int diff = newCapacity - this.items.length;
+            System.arraycopy(newItems, this.head, newItems, this.head + diff, this.items.length - this.head);
+            int idx = this.head;
+            this.head += diff;
+            while (idx < this.head) {
+                newItems[idx] = null;
+                idx++;
+            }
+        }
+
+        this.items = newItems;
+    }
+
+    @Override
+    public String toString() {
+        return String.format("Size = %s, Capacity = %s, Head = %s, Tail = %s", size(), this.items.length, this.head, this.tail);
+    }
+
+    //endregion
+
+    //region WrapQueue
+
+    @RequiredArgsConstructor
+    @NotThreadSafe
+    private static class WrapQueue<T> extends AbstractCollection<T> implements Queue<T> {
+        private final Object[] items;
+        private int nextIndex = 0;
+
+        //region Queue Implementation
+
+        @Override
+        @SuppressWarnings("unchecked")
+        public T poll() {
+            if (this.nextIndex >= this.items.length) {
+                return null;
+            }
+
+            return (T) this.items[this.nextIndex++];
+        }
+
+        @Override
+        public T element() {
+            T result = poll();
+            if (result == null) {
+                throw new NoSuchElementException();
+            }
+
+            return result;
+        }
+
+        @Override
+        @SuppressWarnings("unchecked")
+        public T peek() {
+            return this.nextIndex >= this.items.length ? null : (T) this.items[nextIndex];
+        }
+
+        @Override
+        @SuppressWarnings("unchecked")
+        public Iterator<T> iterator() {
+            return Arrays.stream(this.items).skip(this.nextIndex).map(o -> (T) o).iterator();
+        }
+
+        @Override
+        public int size() {
+            return this.items.length - this.nextIndex;
+        }
+
+        //endregion
+
+        //region Unsupported Methods
+
+        @Override
+        public boolean offer(T t) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public boolean remove(Object o) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public T remove() {
```

Comment:
Here `remove` is unimplemented but the function `element` above is doing what remove is supposed to be doing.

Reply:
Fixed.